### PR TITLE
Add indexing support to data-dictionaries

### PR DESCRIPTION
**Pull request by clayliddell**
fixes WCMS-11068

Add index support to data-dictionaries in DKAN.

- [X] Test coverage exists
- [ ] Documentation exists

## QA Steps

- [x] Create a data-dictionary with different field types (date, string, integer) and indexes for fields of those types.
- [x] Configure the data-dictionary to be the global data-dictionary.
- [x] Create a dataset with a CSV file containing those fields.
- [x] Run cron.
- [x] Verify the indexes were created through a MySQL terminal connection (you can use `SHOW INDEXES FROM table_name`).

### CODE REVIEW

File: modules/json_form_widget/src/ArrayHelper.php [hunk 1]
```diff
@@ -50,118 +50,191 @@ public function __construct(ObjectHelper $object_helper) {
...
+    // Initialize target element to root form render array.
+    $target_element = $form;
+    // Iterate down element heritage from root form element in order to find
+    // immediate parent wrapper element.
+    foreach ($button_heritage as $button_ancestor) {
+      // Navigate deeper into form hierarchy according to the next listed
+      // button field ancestor.
+      $target_element = $target_element[$button_ancestor];
+      if ($button_ancestor === $button_parent) {
+        // We've found the parent element, so we can break early and return it.
```
Comment (clayliddell) [inline modules/json_form_widget/src/ArrayHelper.php]:
```suggestion
        // We've found the parent element, so we can return it.
```

**Comment by janette** [comment]
Trying to index a field that is not defined in the dictionary:
```
TypeError: Argument 4 passed to Drupal\datastore\DataDictionary\AlterTableQuery\MySQLQuery::buildIndexFieldOption() must be of the type string, null given, called in /var/www/dkan/modules/datastore/src/DataDictionary/AlterTableQuery/MySQLQuery.php on line 328 in Drupal\datastore\DataDictionary\AlterTableQuery\MySQLQuery->buildIndexFieldOption() (line 374 of /var/www/dkan/modules/datastore/src/DataDictionary/AlterTableQuery/MySQLQuery.php).
```

**Review by janette — DISMISSED** [review]

**Comment by janette** [comment]
```
1) Drupal\Tests\datastore\Functional\DictionaryEnforcerTest::testDictionaryEnforcement
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    'numOfColumns' => 5
+    'numOfColumns' => 4
     'columns' => Array (
         'record_number' => Array (...)
         'a' => Array (...)
         'b' => Array (...)
         'c' => Array (...)
-        'd' => Array (...)
     )
-    'indexes' => Array (...)
-    'fulltextIndexes' => Array (...)
     'numOfRows' => 2
 )
```

**Review by janette — APPROVED** [review]
Failing test was due to a stale local file. Tests pass once the old file is removed.

**Pull request merged**